bin/generate-phpstorm-stubs-map.php

```diff
@@ -0,0 +1,144 @@
+        }
+
+        if ($node instanceof Node\Stmt\Function_) {
+            /** @psalm-suppress UndefinedPropertyFetch */
+            $this->functionNames[] = $node->namespacedName->toString();
+
+            return NodeTraverser::DONT_TRAVERSE_CHILDREN;
+        }
+
+        return null;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getClassNames() : array
+    {
+        return $this->classNames;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getFunctionNames() : array
+    {
+        return $this->functionNames;
+    }
+
+    public function clear() : void
+    {
+        $this->classNames    = [];
+        $this->functionNames = [];
+    }
+};
+
+$nodeTraverser = new NodeTraverser();
+$nodeTraverser->addVisitor(new NameResolver());
+$nodeTraverser->addVisitor($fileVisitor);
+
+$map = ['classes' => [], 'functions' => []];
+
+foreach (new DirectoryIterator($phpStormStubsDirectory) as $directoryInfo) {
+    if ($directoryInfo->isDot()) {
+        continue;
+    }
+
+    if (! $directoryInfo->isDir()) {
+        continue;
+    }
+
+    if (in_array($directoryInfo->getBasename(), ['tests', 'meta'], true)) {
+        continue;
+    }
+
+    foreach (new DirectoryIterator($directoryInfo->getPathName()) as $fileInfo) {
+        if ($fileInfo->isDot()) {
+            continue;
+        }
+
+        if (! preg_match('/\.php$/', $fileInfo->getBasename())) {
+            continue;
+        }
+
+        $ast = $phpParser->parse(file_get_contents($fileInfo->getPathname()));
+
+        $nodeTraverser->traverse($ast);
+
+        /** @psalm-suppress UndefinedMethod */
+        foreach ($fileVisitor->getClassNames() as $className) {
+            $map['classes'][$className] = $fileInfo->getPathname();
+        }
+
+        /** @psalm-suppress UndefinedMethod */
+        foreach ($fileVisitor->getFunctionNames() as $functionName) {
+            $map['functions'][$functionName] = $fileInfo->getPathname();
+        }
+
+        $fileVisitor->clear();
+    }
+}
+
+$uniqueFiles = [];
+foreach ($map as $typeFiles) {
+    foreach ($typeFiles as $file) {
+        $uniqueFiles[$file] = null;
+    }
+}
+
+ksort($uniqueFiles);
+
+$fileNumbers = array_flip(array_keys($uniqueFiles));
+
+$output = "<?php\n\ndeclare(strict_types=1);\n\n\$files = [\n";
+foreach ($uniqueFiles as $fileName => $tmp) {
+    $relativeFileName = str_replace('\\', '/', substr($fileName, strlen($phpStormStubsDirectory) + 1));
+
+    $output .= sprintf('    \'%s\',%s', $relativeFileName, "\n");
+}
+$output .= "];\n\n";
+
+$output .= "return [\n";
+foreach ($map as $typeFiles) {
+    $output .= "    [\n";
+
+    ksort($typeFiles);
+    foreach ($typeFiles as $name => $file) {
+        $output .= sprintf('        \'%s\' => &$files[%s],%s', $name, $fileNumbers[$file], "\n");
+    }
+
+    $output .= "    ],\n";
+}
+$output .= "];\n";
+
+file_put_contents(__DIR__ . '/../src/SourceLocator/SourceStubber/PhpStormStubsMap.php', $output);
```

Comment:
Same here: the write might fail

Reply:
An assertion here is indeed required :+1: